tests/BatchTest.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Stancl\Tenancy\Tests;
+
+use Illuminate\Bus\BatchRepository;
+use Illuminate\Support\Facades\Event;
+use Illuminate\Support\Str;
+use ReflectionClass;
+use Stancl\Tenancy\Bootstrappers\BatchTenancyBootstrapper;
+use Stancl\Tenancy\Bootstrappers\DatabaseTenancyBootstrapper;
+use Stancl\Tenancy\Events\TenancyEnded;
+use Stancl\Tenancy\Events\TenancyInitialized;
+use Stancl\Tenancy\Listeners\BootstrapTenancy;
+use Stancl\Tenancy\Listeners\RevertToCentralContext;
+use Stancl\Tenancy\Tests\Etc\Tenant;
+
+class BatchTest extends TestCase
+{
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->app->singleton(BatchTenancyBootstrapper::class);
```

Comment:
I don't think tests should be doing this, the package code should take care of it

Reply:
The test failed when I removed this line. 
<img width="1139" alt="image" src="https://user-images.githubusercontent.com/54532330/186656997-8623e3a9-ebbe-443f-a403-3e1e883d522c.png">

 But, strangely, only the last assertion failed.

Reply:
See if other bootstrappers are ever registered as singletons (in the service provider). As I said, it *may* be correct to register it as a singleton, but it's likely not correct to do it in a test.

Reply:
Bootstrappers are registered as singletons: https://github.com/archtechx/tenancy/blob/3.x/src/TenancyServiceProvider.php#L53

Therefore registering the singleton in the test again is unnecessary.

So the issue is likely that the bootstrapper wasn't added to the config early enough in the lifecycle?

Reply:
> So the issue is likely that the bootstrapper wasn't added to the config early enough in the lifecycle?

I think the issue is something else. 

> Bootstrappers are registered as singletons: [3.x/src/TenancyServiceProvider.php#L53](https://github.com/archtechx/tenancy/blob/3.x/src/TenancyServiceProvider.php?rgh-link-date=2022-08-25T17%3A45%3A32Z#L53)

Yes, Bootstrappers are registered. But I noticed TenancyServiceProvider picking config from assets/config. Here I dump the Bootstrappers in the register loop. 
<img width="539" alt="image" src="https://user-images.githubusercontent.com/54532330/186899357-c2f43884-7723-42d3-894e-ee9df8cf5397.png">

These Bootstrappers are always the same doesn't matter what we define as a config in tests. 

As RedisBootstrapper is commented, you are manually registering it in TestCase, so Redis tests would pass. https://github.com/archtechx/tenancy/blob/79e3d53b06f33aa6e30ee4454177e1d798918704/tests/TestCase.php#L108

So either we should enable it by default ([I see you asked](https://github.com/archtechx/tenancy/pull/874#discussion_r955257750)) or register it in TestCase.

Reply:
> enable it by default

We can also just enable it in the test.

Reply:
> These Bootstrappers are always the same doesn't matter what we define as a config in tests

You can set config values and it will work, it just has to be done at the right point in the lifecycle. If setting `config(['foo' => 'bar'])` doesn't work somewhere, a different hook needs to be used.

But I'm 99% sure that using `config(['foo' => 'bar'])` at the start of the test or in `setUp()` will work.

Reply:
> But I'm 99% sure that using config(['foo' => 'bar']) at the start of the test or in setUp() will work.

I tried both. I set the config in setUp method and also at the start of the tests. Nothing works. Maybe I'm missing something but when I dump in the singleton registration, and in tests that how the flow looks like. 

<img width="499" alt="image" src="https://user-images.githubusercontent.com/54532330/187653634-3663b567-e72b-4baf-b68b-c6a5b0b77c31.png">

You can see, all the singleton above. The setUp dump call is in `setUp` method. It's being executed after the singletons got registered.

Reply:
As discussed internally, resolving this by registering the singleton in TestCase with todo to find a nice way afterward.